File: src/main/java/net/sf/rails/game/state/WalletBag.java
package net.sf.rails.game.state;

import com.google.common.base.Preconditions;

/**
 * A WalletBag allows the storage of only one item of the specified class
 */

public class WalletBag<T extends Countable> extends Wallet<T> {
    
    private final T item;
    private int amount = 0;
    
    private WalletBag(Owner parent, String id, Class<T> type, T item) {
        super(parent, id, type);
        this.item = item;
    }
    
    /**
     * Creates an empty WalletBag
     */
    public static <T extends Countable> WalletBag<T> create(Owner parent, String id, Class<T> type, T item){
        return new WalletBag<T>(parent, id, type, item);
    }
    
    /**
     * @param item for which the value is retrieved
     * @return the current amount of the item inside the wallet
     */
    @Override
    public int value(T item) {
        Preconditions.checkArgument(item == this.item, "WalletBag only accepts item " + this.item);
        return amount;
    }

    @Override
    public int value() {
        return amount;
    }
    
    @Override
    void change (T item, int value) {
        amount += value;
    }

    @Override
    public String toText() {
        return Integer.toString(amount);
    }

}

File: dotnet/RailsLib.Net/Net/Game/State/WalletBag.cs
using GameLib.Net.Util;
using System;
using System.Collections.Generic;
using System.Text;

/**
 * A WalletBag allows the storage of only one item of the specified class
 */

namespace GameLib.Net.Game.State
{
    public class WalletBag<T> : Wallet<T> where T : ICountable
    {
        private T item;
        private int amount = 0;

        private WalletBag(IOwner parent, string id, T item) : base(parent, id)
        {
            this.item = item;
        }

        /**
         * Creates an empty WalletBag
         */
        public static WalletBag<T> Create(IOwner parent, string id, T item)
        {
            return new WalletBag<T>(parent, id, item);
        }

        /**
         * @param item for which the value is retrieved
         * @return the current amount of the item inside the wallet
         */
        override public int Value(T item)
        {
            Precondition.CheckArgument(item.Equals(this.item), "WalletBag only accepts item " + this.item);
            return amount;
        }

        override public int Value()
        {
            return amount;
        }

        override public void Change(T item, int value)
        {
            amount += value;
        }


        override public string ToText()
        {
            return amount.ToString();
        }

    }
}
